vendor/github.com/danwinship/knftables/types.go

```diff
@@ -0,0 +1,338 @@
+	// regular chain.
+	Type *BaseChainType
+	// Hook is the hook that the chain is connected to; this must be set for a base
+	// chain and unset for a regular chain.
+	Hook *BaseChainHook
+	// Priority is the chain priority; this must be set for a base chain and unset for
+	// a regular chain. You can call ParsePriority() to convert this to a number.
+	Priority *BaseChainPriority
+
+	// Comment is an optional comment for the object.
+	Comment *string
+
+	// Handle is an identifier that can be used to uniquely identify an object when
+	// deleting it. When adding a new object, this must be nil
+	Handle *int
+}
+
+// Rule represents a rule in a chain
+type Rule struct {
+	// Chain is the name of the chain that contains this rule
+	Chain string
+
+	// Rule is the rule in standard nftables syntax. (Should be empty on Delete, but
+	// is ignored if not.) Note that this does not include any rule comment, which is
+	// separate from the rule itself.
+	Rule string
+
+	// Comment is an optional comment for the rule.
+	Comment *string
+
+	// Index is the number of a rule (counting from 0) to Add this Rule after or
+	// Insert it before. Cannot be specified along with Handle. If neither Index
+	// nor Handle is specified then Add appends the rule the end of the chain and
+	// Insert prepends it to the beginning.
+	Index *int
+
+	// Handle is a rule handle. In Add or Insert, if set, this is the handle of
+	// existing rule to put the new rule after/before. In Delete or Replace, this
+	// indicates the existing rule to delete/replace, and is mandatory. In the result
+	// of a List, this will indicate the rule's handle that can then be used in a
+	// later operation.
+	Handle *int
+}
+
+// SetFlag represents a set or map flag
+type SetFlag string
+
+const (
+	// ConstantFlag is a flag indicating that the set/map is constant. FIXME UNDOCUMENTED
+	ConstantFlag SetFlag = "constant"
+
+	// DynamicFlag is a flag indicating that the set contains stateful objects
+	// (counters, quotas, or limits) that will be dynamically updated.
+	DynamicFlag SetFlag = "dynamic"
+
+	// IntervalFlag is a flag indicating that the set contains either CIDR elements or
+	// IP ranges.
+	IntervalFlag SetFlag = "interval"
+
+	// TimeoutFlag is a flag indicating that the set/map has a timeout after which
+	// dynamically added elements will be removed. (It is set automatically if the
+	// set/map has a Timeout.)
+	TimeoutFlag SetFlag = "timeout"
+)
+
+// SetPolicy represents a set or map storage policy
+type SetPolicy string
+
+const (
+	// PolicyPerformance FIXME
+	PerformancePolicy SetPolicy = "performance"
+
+	// PolicyMemory FIXME
+	MemoryPolicy SetPolicy = "memory"
+)
+
+// Set represents the definition of an nftables set (but not its elements)
+type Set struct {
+	// Name is the name of the set.
+	Name string
+
+	// Type is the type of the set key (eg "ipv4_addr"). Either Type or TypeOf, but
+	// not both, must be non-empty.
+	Type string
```

Comment:
And while I'm pointing out knftables API bits I'm not sure about...

The nftables syntax looks like:

```
add set ip kube-proxy firewall { type ipv4_addr . inet_proto . inet_service }
add element ip kube-proxy firewall { 10.1.2.3 . tcp . 80 }
```

But right now in the knftables API, `Element` has `Key []string`, but `Set` has `Type string`, so you say:

```
tx.Add(&knftables.Set{
        Name: "firewall",
        Type: "ipv4_addr . inet_proto . inet_service",
})
tx.Add(&knftables.Element{
        Set: "firewall",
        Key: []string{"10.1.2.3", "tcp", "80"},
})
```

I thought about changing `Set.Type` to be `[]string` too, for consistency (and so we're not exposing too much nftables syntax in the API), but then this causes a problem with `Map`, because currently `Map.Type` includes both the key type and the value type (e.g. `Type: "ipv4_addr  . inet_proto . inet_service : verdict"`), so we'd have to split that into `KeyType []string, ValueType []string`... But maybe that would be better.

The other tricky part with splitting `Type` is that we have to support `TypeOf` as well (there are certain kinds of values that don't have a named type, so the only way to use them in a set/map key/value is to say "use the type of this expression"). So we'd need `KeyType`, `ValueType`, `KeyTypeOf`, and `ValueTypeOf`, and validate that if you use the type-of variant for one, that you also used it for the other. Or change the way that we handle `type` vs `typeof`, and just have `TypeOf bool` to indicate "use typeof rather than type"?

Reply:
Related: it annoys me that I can do:
```
        Rule: knftables.Concat(
                "ip", "daddr", clusterIP,
                "meta", "l4proto", protocol,
                "th", "dport", svcInfo.Port(),
        ),
```
in a rule, but in an element, I have to do
```
        Key: []string{
                clusterIP.String(),
                protocol,
                strconv.Itoa(svcInfo.Port())
        },
```

Reply:
I think that 90% of this feature is the nftables API, the proxier implementation is not something that worries me since is translating the proxier logic to use the nftables "magic" that solves the existing iptables bottleneck, and I think we have a good coverage on testing for it, with the addition that is alpha gated and is really safe to keep iterating.

Is the nft API that worries me, is a direct dependency and is external, I don't trust any userspace library to not have bugs or break compatibility because of reasons, we hit this before https://github.com/kubernetes/kubernetes/issues/112477#issuecomment-1248660064

If this happens and we need to backport any change that means revendoring an entire external project with a new release, with the risks of regressions ... also, since is developed out of tree https://github.com/danwinship/knftables/commits/master we'll not have signal if is compatible until we try to update it.

I think the nft wrapper  should go intree with the proxier ... it can be published externally by using staging/kube-proxy if desired

Reply:
> I think the nft wrapper should go intree with the proxier ... it can be published externally by using staging/kube-proxy if desired

So I know that you and I have had this conversation, but for everyone just joining...

The reason the nftables library is external is that it is _not_ intended to be only/especially for kube-proxy. I already have draft PRs against [CNI](https://github.com/containernetworking/plugins/pull/935) and [ovn-kubernetes](https://github.com/ovn-org/ovn-kubernetes/pull/3709) (which uses iptables for a handful of miscellaneous rules), and plan to get to [cri-o](https://github.com/cri-o/cri-o/blob/v1.28.0/internal/hostport/hostport.go#L135), the non-kube-proxy part of [openshift-sdn](https://github.com/openshift/sdn/blob/release-4.14/pkg/network/node/iptables.go) and a few other more-OpenShift-specific things. We're planning to eventually write a NetworkPolicy implementation using nftables as well. And that's just what _I'm_ working on. My expectation is that this library will be used by other people as well.

> I don't trust any userspace library to not have bugs or break compatibility because of reasons
> ...
> that means revendoring an entire external project with a new release, with the risks of regressions
> ...
> we'll not have signal if is compatible until we try to update it.

These are all arguments for not vendoring any library ever, and that ship has kinda already sailed.

Reply:
> So I know that you and I have had this conversation, but for everyone just joining...

yeah, I just thought it will be better if this gets documented on the review 

> The reason the nftables library is external is that it is not intended to be only/especially for kube-proxy. I already have draft PRs against https://github.com/containernetworking/plugins/pull/935 and https://github.com/ovn-org/ovn-kubernetes/pull/3709 (which uses iptables for a handful of miscellaneous rules), and plan to get to [cri-o](https://github.com/cri-o/cri-o/blob/v1.28.0/internal/hostport/hostport.go#L135), the non-kube-proxy part of [openshift-sdn](https://github.com/openshift/sdn/blob/release-4.14/pkg/network/node/iptables.go) and a few other more-OpenShift-specific things.

well, that is something that alleviates my concern, as it will be more users,  the library will benefit of more eyeballs ... for me your argument is fair enough and I'm ok with this approach

Reply:
Oh, the other thing is that I do want to get it out of my `danwinship/` at some point, possibly to `kubernetes-sigs`. (While it's not intended only for kubernetes, it is explicitly intended primarily for kubernetes-adjacent things.) Maybe that should happen at about the point where we bump to `v0.1.0` and a stable-for-now API.

Reply:
I'm 100% ok having it in`kubernetes-sigs`, thanks for clarifying

Reply:
is reasonable ask for the library to be ~stable as criteria for beta?

Reply:
definitely

Reply:
I don't have time today to take apart the whole library, but I thank you for calling out places where it's not up to your bar yet.